state-chain/pallets/cf-ingress-egress/src/tests/boost.rs

```diff
@@ -0,0 +1,697 @@
+			assert_boosted(deposit_address, deposit_id, [BoostPoolTier::FiveBps]);
+		}
+	});
+}
+
+#[test]
+fn double_prewitness_due_to_reorg() {
+	new_test_ext().execute_with(|| {
+		const BOOSTER_AMOUNT_1: AssetAmount = 500_000_000;
+		const DEPOSIT_AMOUNT: AssetAmount = 250_000_000;
+		const BOOST_FEE_BPS: BasisPoints = 10;
+		const BOOST_FEE: AssetAmount = DEPOSIT_AMOUNT * BOOST_FEE_BPS as u128 / 10_000;
+
+		setup();
+
+		assert_ok!(IngressEgress::add_boost_funds(
+			RuntimeOrigin::signed(BOOSTER_1),
+			eth::Asset::Eth,
+			BOOSTER_AMOUNT_1,
+			BoostPoolTier::TenBps
+		));
+
+		// ==== LP sends funds to liquidity deposit address, which gets pre-witnessed ====
+		let (channel_id, deposit_address) = request_deposit_address_eth(LP_ACCOUNT, BOOST_FEE_BPS);
+		let deposit_id1 = prewitness_deposit(deposit_address, eth::Asset::Eth, DEPOSIT_AMOUNT);
+
+		const LP_BALANCE_AFTER_BOOST: AssetAmount =
+			INIT_LP_BALANCE + DEPOSIT_AMOUNT - BOOST_FEE - INGRESS_FEE;
+
+		// First deposit should be boosted, crediting the LP as per channel action:
+		{
+			assert_boosted(deposit_address, deposit_id1, [BoostPoolTier::TenBps]);
+
+			assert_eq!(get_lp_eth_balance(&LP_ACCOUNT), LP_BALANCE_AFTER_BOOST);
+
+			assert_eq!(
+				get_available_amount(eth::Asset::Eth, BoostPoolTier::TenBps),
+				BOOSTER_AMOUNT_1 - DEPOSIT_AMOUNT + BOOST_FEE
+			);
+		}
+
+		// Due to reorg, the same deposit is pre-witnessed again, but it has no effect since
+		// we don't boost it due to an existing boost:
+		let deposit_id2 = prewitness_deposit(deposit_address, eth::Asset::Eth, DEPOSIT_AMOUNT);
+		{
+			// No channel action took place, LP balance is unchanged:
+			assert_eq!(get_lp_eth_balance(&LP_ACCOUNT), LP_BALANCE_AFTER_BOOST);
+
+			// No funds from the boost pool are consumed:
+			assert_eq!(
+				get_available_amount(eth::Asset::Eth, BoostPoolTier::TenBps),
+				BOOSTER_AMOUNT_1 - DEPOSIT_AMOUNT + BOOST_FEE
+			);
+		}
+
+		// The deposit is finally fully witnessed, it has no effect on the LP, but
+		// boosters get credited
+		{
+			witness_deposit(deposit_address, eth::Asset::Eth, DEPOSIT_AMOUNT);
+
+			assert_eq!(get_lp_eth_balance(&LP_ACCOUNT), LP_BALANCE_AFTER_BOOST);
+
+			assert_eq!(
+				get_available_amount(eth::Asset::Eth, BoostPoolTier::TenBps),
+				BOOSTER_AMOUNT_1 + BOOST_FEE
+			);
+
+			// One prewitness deposit is "consumed", but one is still pending:
+			assert_eq!(PrewitnessedDeposits::<Test>::get(channel_id, deposit_id1), None);
+			assert!(PrewitnessedDeposits::<Test>::get(channel_id, deposit_id2).is_some());
+		}
+
+		// When the channel expires, the redundant prewitnessed deposit is finally removed:
+		{
+			let recycle_block = IngressEgress::expiry_and_recycle_block_height().2;
+			BlockHeightProvider::<MockEthereum>::set_block_height(recycle_block);
+			IngressEgress::on_idle(recycle_block, Weight::MAX);
+			assert_eq!(PrewitnessedDeposits::<Test>::get(channel_id, deposit_id2), None);
+		}
```

Comment:
I was under the impression that we'd still want `PrewitnessedDeposits` for prewitness subscribtion (in case it's still needed for some reason), but looks like it is not used for that. I think it might be possible to put the boosted amount directly into the Boosted status on the channel and then `PrewitnessedDeposits` won't be needed at all? Might need to think about it a bit more.